# Heart Disease Prediction using Machine Learning

## AIMLCZG523 - Machine Learning Operations (MLOps)

### Objective

This project aims to build an end-to-end Machine Learning pipeline to predict the presence of heart disease using clinical patient data from the UCI Machine Learning Repository.

The project demonstrates the complete MLOps lifecycle:

- Data Acquisition
- Exploratory Data Analysis (EDA)
- Data Preprocessing
- Model Development
- Experiment Tracking (MLflow)
- API Development (FastAPI)
- Containerization (Docker)
- CI/CD
- Deployment
- Monitoring

---

# 1. Dataset Information

## Dataset

- **Name:** Heart Disease Dataset
- **Source:** UCI Machine Learning Repository
- **Dataset ID:** 45

### Problem Statement

Predict whether a patient has heart disease based on clinical attributes such as age, cholesterol, chest pain type, blood pressure, ECG results, and maximum heart rate.

### Project Dataset

A local copy of the dataset has been stored under:

```text
data/raw/heart.csv
```

All subsequent analysis and model development will use this local copy to ensure reproducibility.

In [1]:
# ===========================
# Import Libraries
# ===========================

from pathlib import Path

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

# Notebook Settings
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 10)

plt.style.use("default")

In [2]:
# ===========================
# Load Dataset
# ===========================

PROJECT_ROOT = Path.cwd().parent
DATA_PATH = PROJECT_ROOT / "data" / "raw" / "heart.csv"

df = pd.read_csv(DATA_PATH)

print(f"Dataset loaded successfully from:\n{DATA_PATH}")

Dataset loaded successfully from:
D:\WILP\Semester 3\MLOPs\Heart_Disease_MLOps\data\raw\heart.csv


# 2. Initial Dataset Inspection

## Objective

Before performing exploratory analysis, verify:

- Dataset dimensions
- Feature names
- Data types
- Missing values
- Basic statistics

This step ensures the dataset has been loaded correctly and helps identify potential data quality issues.

In [3]:
# ===========================
# Preview Dataset
# ===========================

df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,num
0,63,1,1,145,233,1,2,150,0,2.3,3,0.0,6.0,0
1,67,1,4,160,286,0,2,108,1,1.5,2,3.0,3.0,2
2,67,1,4,120,229,0,2,129,1,2.6,2,2.0,7.0,1
3,37,1,3,130,250,0,0,187,0,3.5,3,0.0,3.0,0
4,41,0,2,130,204,0,2,172,0,1.4,1,0.0,3.0,0


In [4]:
# ===========================
# Dataset Information
# ===========================

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        299 non-null    float64
 12  thal      301 non-null    float64
 13  num       303 non-null    int64  
dtypes: float64(3), int64(11)
memory usage: 33.3 KB


In [5]:
# ===========================
# Missing Values
# ===========================

missing_values = df.isnull().sum().sort_values(ascending=False)

missing_values

ca          4
thal        2
cp          0
trestbps    0
age         0
           ..
thalach     0
oldpeak     0
exang       0
slope       0
num         0
Length: 14, dtype: int64

In [6]:
# ===========================
# Statistical Summary
# ===========================

df.describe().T

,count,mean,std,min,25%,50%,75%,max
age,303.0,54.438944,9.038662,29.0,48.0,56.0,61.0,77.0
sex,303.0,0.679868,0.467299,0.0,0.0,1.0,1.0,1.0
cp,303.0,3.158416,0.960126,1.0,3.0,3.0,4.0,4.0
trestbps,303.0,131.689769,17.599748,94.0,120.0,130.0,140.0,200.0
chol,303.0,246.693069,51.776918,126.0,211.0,241.0,275.0,564.0
...,...,...,...,...,...,...,...,...
oldpeak,303.0,1.039604,1.161075,0.0,0.0,0.8,1.6,6.2
slope,303.0,1.600660,0.616226,1.0,1.0,2.0,2.0,3.0
ca,299.0,0.672241,0.937438,0.0,0.0,0.0,1.0,3.0
thal,301.0,4.734219,1.939706,3.0,3.0,3.0,7.0,7.0


# 3. Initial Observations

### Dataset Summary

| Property | Observation |
|-----------|-------------|
| Number of Samples | 303 |
| Number of Features | 13 |
| Target Variable | `num` |
| Missing Values | Present in `ca` and `thal` |
| Numerical Features | Majority |
| Categorical Features | Several integer-encoded categorical variables |

### Notes

- The original UCI dataset stores the target as `num` with values from 0 to 4.
- For binary classification, this target will later be transformed into:
  - **0** → No heart disease
  - **1** → Presence of heart disease (values 1–4)

This transformation will be performed during the preprocessing stage, not during exploratory data analysis.